vstruct/defs/pe.py

```diff
@@ -264,11 +265,36 @@ def __init__(self):
         self.FileDateMS         = v_uint32()
         self.FileDateLS         = v_uint32()
 
+class v_name(v_bytes):
+    '''
+    Specifically for things that *should* be utf-8 strings, but aren't really enforced as such
+    '''
+    def __repr__(self):
+        try:
+            name = self._vs_value.decode('utf-8')
+            return name
+        except UnicodeDecodeError:
+            return binascii.hexlify(self._vs_value).decode('utf-8')
```

Comment:
I don't like this implicit conversion to another data format. I'm worried that users will take the string provided here and understand it to be the section name, verbatim. For example, I can imagine an analysis system that extracts PE features and displays them to users; with this change, there is no way for users to know if they can take the section name and create a, e.g. yara rule with it.

Also, what happens when the section name is literally `65656565`?

I believe the best way to handle this is to use `str` for successfully decoded UTF-8 and `bytes` for unrecognized data. Then, users can rely on the str data being correct, and when they encounter bytes, will be forced to handle the edge case in the correct, domain-specific way. Sure, it may surprise a user and may require a patch to their software, but with clear documentation on the field in viv, its a quick fix and leads to correct behavior.

I can also imagine adding a flag to this class that specifies "is this invalid data encoded as hex?". I don't like this solution because its too easy to ignore the flag and render misleading data. Better than this would be to change the repr to be something like `(invalid data) 65656565`, but again, users that blindly fetch via `vsGetValue` and e.g. create a yara rule will be doing this wrong.

Maybe `vsGetValue` should return `Union[str, bytes]` and `repr` should `.text`/`(invalid data) 65656565`.

Reply:
Also, I'd consider this a breaking change that really needs to be explicitly called out in the next release.

Reply:
So, from my perspective, I don't want to change the return type for this to be directly inconsistent with how vsGetValue works for all the other vstruct construct,s and then also have to scatter `isinstance` calls around to deal with the mixed return types. I generally try to avoid mixed return types when I can to keep the API relatively easier to maintain on my part.

But yes, you're right in that right now it's a little too "magic" and not entirely clear when it's valid/when it's not. I'm a lot more fine with the idea of changing the repr/vsGetValue (well, really the repr can be removed now. vsGetValue will almost always be hit) to something like `[invalid name] <unadorned bytes>` along with a flag/`isNameValid()` API on the IMAGE_SECTION_HEADER class for whether or not the section name is valid. I think that hits the best of both worlds of consistency and explicitness and fixes the issue in #499 and while it is an API update, it's not as breaking as it could be.

Thoughts? I feel like this meets us in the middle somewhere, but I'm open to suggestions.

Reply:
> I don't want to change the return type for this to be directly inconsistent with how vsGetValue works for all the other vstruct construct,s and then also have to scatter isinstance calls around to deal with the mixed return types.

I understand this, though I still think using `isinstance` is the *correct* way to do this; there's an edge case with malformed data, so Pythons (weak) type system can help highlight issues for once. Its more convenient to not have to do this, sure. I guess the question is whether viv provides a sane default that works for most, or an implementation that encourages users to consider edge cases that we see in the real world.

Anyways, I get your position, I have a different one, and its ok to disagree. I like the compromise that you propose much better - thanks for suggesting it!

Reply:
Aye. That was the intention. I recognize your way of using vivisect is very different from my own, and I want to make sure things work for both of us.

So if you're cool with the compromise, I'll switch that up later tonight when I get a spot of free time. Thanks for the input by the way! It helps. Like a lot.

Reply:
send it!

Reply:
you both have explained my quandary on this one very well.

1) how to keep it from breaking vstruct is my chief concern :)
2) how to make it relevant for the use-cases is important.  if it doesn't do what we want, why are we developing it in the first place?!
3) how to make the API usable and consistent is also important.  if we make the API too complicated, what unforeseen things will it break going forward?  and that plays into concern #2.